### PR TITLE
Extend child documents model and add HOJKS

### PR DESCRIPTION
**Pull request by Joosakur**

### CODE REVIEW

File: service/src/main/resources/db/migration/V353__child_document_extensions.sql [hunk 1]
```diff
@@ -0,0 +1,21 @@
+ALTER TYPE document_template_type ADD VALUE 'HOJKS';
+
+CREATE TYPE child_document_status AS ENUM (
+    'DRAFT',
+    'PREPARED',
+    'COMPLETED'
+);
+
+ALTER TABLE child_document
+    ADD COLUMN status child_document_status NOT NULL DEFAULT 'DRAFT',
```
Comment (Joosakur) [inline service/src/main/resources/db/migration/V353__child_document_extensions.sql:10]: drop default?

File: service/src/main/resources/db/migration/V353__child_document_extensions.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+UPDATE child_document
+SET status = 'COMPLETED', published_content = content
+WHERE published_at IS NOT NULL;
+
+ALTER TABLE child_document
+    ADD CONSTRAINT published_consistency CHECK ( (published_at IS NULL) = (published_content IS NULL) ),
+    ADD CONSTRAINT non_drafts_are_published CHECK ( status = 'DRAFT' OR published_at IS NOT NULL );
+
+CREATE INDEX idx$child_document_status_unpublished ON child_document(status, (published_at IS NULL));
```
Comment (Joosakur) [inline service/src/main/resources/db/migration/V353__child_document_extensions.sql:21]: child_id index?

File: service/src/main/kotlin/fi/espoo/evaka/shared/security/actionrule/HasGlobalRole.kt [hunk 3]
```diff
@@ -184,4 +186,16 @@ ${
...
+        rule<ChildDocumentId> { _, _ ->
+            sql(
+                """
+SELECT id
+FROM child_document
+WHERE status = ANY(${bind(statuses)}::child_document_status[])
+            """
+                    .trimIndent()
+            )
+        }
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/shared/security/actionrule/HasGlobalRole.kt:200]: remove

File: service/src/main/kotlin/fi/espoo/evaka/shared/dev/DataInitializers.kt [hunk 4]
```diff
@@ -1561,8 +1561,8 @@ fun Database.Transaction.insertTestChildDocument(row: DevChildDocument) =
     insertTestDataRow(
             row,
             """
-INSERT INTO child_document (id, child_id, template_id, content, published_at)
-VALUES (:id, :childId, :templateId, :content, :publishedAt)
+INSERT INTO child_document (id, status, child_id, template_id, content, published_content, published_at)
+VALUES (:id, :status, :childId, :templateId, :content, ${if(row.publishedAt == null) "NULL" else ":content"}, :publishedAt)
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/shared/dev/DataInitializers.kt:1565]: add published content to input

File: service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentQueries.kt [hunk 5]
```diff
@@ -130,34 +168,84 @@ fun Database.Transaction.publishChildDocument(id: ChildDocumentId, now: Helsinki
...
+    documentId: ChildDocumentId,
+    requestedStatus: DocumentStatus,
+    goingForward: Boolean // false = backwards
+): StatusTransition {
+    val document = tx.getChildDocument(documentId) ?: throw NotFound()
+
+    val statusList = document.template.type.states
+    val currentIndex = statusList.indexOf(document.status)
+    if (currentIndex < 0) {
+        throw Error("document $documentId is in invalid status")
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentQueries.kt:201]: error()

File: service/src/main/kotlin/fi/espoo/evaka/document/DocumentTemplate.kt [hunk 6]
```diff
@@ -74,9 +76,11 @@ data class Section(
 
 @Json data class DocumentTemplateContent(val sections: List<Section>)
 
-enum class DocumentType {
-    PEDAGOGICAL_REPORT,
-    PEDAGOGICAL_ASSESSMENT
+@ConstList("documentTypes")
+enum class DocumentType(val states: List<DocumentStatus>) {
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/document/DocumentTemplate.kt:80]: comment

File: service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentController.kt [hunk 7]
```diff
@@ -193,26 +205,76 @@ class ChildDocumentController(
...
+        @PathVariable documentId: ChildDocumentId,
+        @RequestBody body: StatusChangeRequest
+    ) {
+        db.connect { dbc ->
+                dbc.transaction { tx ->
+                    accessControl.requirePermissionFor(
+                        tx,
+                        user,
+                        clock,
+                        Action.ChildDocument.NEXT_STATE,
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentController.kt:262]: prev_state

File: service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentController.kt [hunk 8]
```diff
@@ -231,11 +293,8 @@ class ChildDocumentController(
                         Action.ChildDocument.DELETE,
                         documentId
                     )
-                    tx.getChildDocument(documentId)?.also {
-                        if (it.publishedAt != null)
-                            throw BadRequest("Can not delete published document")
-                    }
-                    tx.deleteChildDocumentDraft(documentId)
+                    tx.getChildDocument(documentId)
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentController.kt:296]: remove?

File: service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentController.kt [hunk 9]
```diff
@@ -231,11 +293,8 @@ class ChildDocumentController(
...
+                    tx.deleteUnpublishedChildDocumentDraft(documentId)
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentController.kt:297]: not unpublished

File: service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentQueries.kt (hunk 5 above)
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentQueries.kt:201]: error()

File: service/src/integrationTest/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentControllerCitizenIntegrationTest.kt [hunk 10]
```diff
@@ -170,4 +179,55 @@ class ChildDocumentControllerCitizenIntegrationTest :
...
         )
     }
+
+    @Test
+    fun `Updates after publish are shown only after republish`() {
+        employeeController.publishDocument(dbInstance(), employeeUser, clock, documentId)
+
+        assertEquals(
+            mapOf(testChild_1.id to 1),
+            controller.getUnreadCount(dbInstance(), citizen, clock)
```
Comment (Joosakur) [inline service/src/integrationTest/kotlin/fi/espoo/evaka/document/childdocument/ChildDocumentControllerCitizenIntegrationTest.kt:189]: wrap

File: frontend/src/lib-customizations/defaults/citizen/i18n/sv.tsx [hunk 11]
```diff
@@ -2467,8 +2467,11 @@ const sv: Translations = {
     },
     vasu: {
       title: 'Pedagogiska dokument',
-      plansTitle: 'Planer',
+      plansTitle:
+        'Planer för småbarnspedagogik och lärande inom förskoleundervisning',
       noVasus: 'Inga planer',
+      hojksTitle:
+        'Henkilökohtaisen opetuksen järjestämistä koskevat suunnitelmat', // TODO
```
Comment (Joosakur) [inline frontend/src/lib-customizations/defaults/citizen/i18n/sv.tsx:2474]: käännä

File: frontend/src/lib-components/molecules/ConfirmedMutation.tsx [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+          rejectLabel={cancelLabel ?? i18n.common.cancel}
+        />
+      )}
+    </div>
+  )
+}
+
+export const ConfirmedMutation = React.memo(
+  ConfirmedMutation_
+) as typeof ConfirmedMutation_
```
Comment (Joosakur) [inline frontend/src/lib-components/molecules/ConfirmedMutation.tsx:77]: why _?

File: frontend/src/employee-frontend/components/child-documents/ChildDocumentEditor.tsx [hunk 13]
```diff
@@ -217,22 +238,69 @@ const ChildDocumentEditorView = React.memo(function ChildDocumentEditorView({
...
+              {!preview && (
+                <Button
+                  text={i18n.childInformation.childDocuments.editor.preview}
+                  primary
+                  onClick={() => setPreview(true)}
+                  disabled={!saved}
+                  data-qa="preview-button"
+                />
+              )}
+              {preview && permittedActions.includes('PUBLISH') && (
```
Comment (Joosakur) [inline frontend/src/employee-frontend/components/child-documents/ChildDocumentEditor.tsx:258]: != 'completed'

File: frontend/src/citizen-frontend/children/sections/vasu-and-leops/VasuAndLeopsSection.tsx [hunk 14]
```diff
@@ -244,8 +252,19 @@ export default React.memo(function PedagogicalDocumentsSection({
...
+            {featureFlags.experimental?.hojks && (
+              <>
+                <H3>{i18n.children.vasu.hojksTitle}</H3>
+                <ChildDocumentsContent childId={childId} types={['HOJKS']} />
+              </>
+            )}
+            <H3>{i18n.children.vasu.otherDocumentsTitle}</H3>
+            <ChildDocumentsContent
+              childId={childId}
+              types={['PEDAGOGICAL_ASSESSMENT', 'PEDAGOGICAL_REPORT']}
```
Comment (Joosakur) [inline frontend/src/citizen-frontend/children/sections/vasu-and-leops/VasuAndLeopsSection.tsx:266]: valueList.filter

File: service/src/main/resources/db/migration/V353__child_document_extensions.sql (hunk 2 above)
Comment (Joosakur) [reply service/src/main/resources/db/migration/V353__child_document_extensions.sql]: already exists

**Review by Gekkio — APPROVED** [review]

**Pull request merged**